### DISCUSSION TITLE
The requested module '@mikro-orm/postgresql' does not provide an export named 'Opt'

### DISCUSSION THREAD

**Discussion by linkthai** [mikro-orm/mikro-orm#5097] [General]
### Describe the bug

MikroORM version: 6.0.1
OS: MacOS
Node: 20

Here is the file that cause problem:
``` ts
import { Entity, Index, PrimaryKey, Property, Opt } from '@mikro-orm/postgresql';

@Entity({ abstract: true })
export class Base {
  static entityName: string;
  static entitiesName: string;

  @PrimaryKey({ type: 'uuid', defaultRaw: 'gen_random_uuid()' })
  id: Opt<string>;

  @Index()
  @Property({ type: 'datetime', onCreate: () => new Date() })
  createdAt: Opt<Date>;

  @Property({
    type: 'datetime',
    onCreate: () => new Date(),
    onUpdate: () => new Date(),
  })
  updatedAt: Opt<Date>;
}
```

Running `mikro-orm schema` or `mikro-orm migration` will cause the following error 
``` ts
import { Entity, Index, PrimaryKey, Property, Opt } from '@mikro-orm/postgresql';
                                                                     ^^^
SyntaxError: The requested module '@mikro-orm/postgresql' does not provide an export named 'Opt'
```

I fixed it by moving Opt to a separate type definition
``` ts
import type { Opt } from '@mikro-orm/postgresql';
import { Entity, Index, PrimaryKey, Property } from '@mikro-orm/postgresql';
```

Note: This is my custom loader to run MikroORM in Typescript so it could be the source of the problem.
``` js
// loader.js
// https://github.com/TypeStrong/ts-node/discussions/1450
import { resolve as resolveTs } from 'ts-node/esm';
import * as tsConfigPaths from 'tsconfig-paths';
import { pathToFileURL } from 'url';

const { absoluteBaseUrl, paths } = tsConfigPaths.loadConfig();
const matchPath = tsConfigPaths.createMatchPath(absoluteBaseUrl, paths);

export function resolve(specifier, ctx, defaultResolve) {
  const match = matchPath(specifier);
  return match
    ? resolveTs(pathToFileURL(`${match}`).href, ctx, defaultResolve)
    : resolveTs(specifier, ctx, defaultResolve);
}

export { load, transformSource } from 'ts-node/esm';
```

My schema script 
``` json
"schema": "node --no-warnings --loader ./loader.js ./node_modules/@mikro-orm/cli/cli.js schema:update",
```

MikroORM.config.ts
``` ts
export default defineConfig({
  entities: ['./src/modules/**/*.entity.ts'],
  entitiesTs: ['./src/modules/**/*.entity.ts'],
  metadataProvider: ReflectMetadataProvider,
  host: process.env.DATABASE_HOST,
  port: Number(process.env.DATABASE_PORT || 5432),
  user: process.env.DATABASE_USERNAME,
  password: process.env.DATABASE_PASSWORD,
  dbName: process.env.DATABASE_NAME,
  forceUtcTimezone: true,
  ...(!(process.env.DATABASE_HOST as string).includes('localhost')
    ? {
        driverOptions: {
          connection: {
            ssl: {
              ca: fs.readFileSync('global-bundle.pem'),
            },
          },
        },
      }
    : {}),
  schemaGenerator: {
    disableForeignKeys: false,
  },
  migrations: {
    path: './src/migrations',
    transactional: true,
    allOrNothing: true,
    disableForeignKeys: false,
  },
  seeder: {
    path: 'dist/seeders',
    pathTs: 'seeders',
    defaultSeeder: 'DatabaseSeeder',
    fileName: (className: string) =>
      `${className
        .replace('Seeder', '')
        .replace(
          /[A-Z]/g,
          (match, offset) => `${offset !== 0 ? '-' : ''}${match.toLowerCase()}`,
        )}.seeder`,
  },
  extensions: [SeedManager, Migrator],
});
```

### Reproduction

[Code Sandbox](https://codesandbox.io/p/github/linkthai/opt-error/master
)
I can't reproduce it programmatically, it's only CLI related, it could be because of my own way of running the CLI as decribed above. In case you wouldn't find a hint of how this happens, I'll close the issue as it a mild inconvinience with an easy fix at worst that only happens to my setup.

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.0.1

### Node.js version

20.5.1

### Operating system

MacOS

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
> import type { Opt } from '@mikro-orm/postgresql';
> import { Entity, Index, PrimaryKey, Property } from '@mikro-orm/postgresql';

It's a type (and it indeed is exported from there), so doing this feels like a proper fix. I recall some issues about `Ref` type with the very same symptoms, so it's "nothing new" (not caused by v6 changes). Could be about the ESM support, as we don't compile the full project, just ESM named exports via `gen-esm-wrapper`.

> it could be because of my own way of running the CLI as decribed above

Will test your loader to see if it breaks on my end, it's indeed suspicious.

>   id: Opt<string>;

You don't need to mark `id` property with it, that's implicit (same for `_id` and `uuid` named properties).

**Reply by linkthai** [reply]
Thanks for the quick reply. I actually just encountered another strange problem.

I have two apps that just upgraded to 6.0.1. The first one is fine except for the `Opt` problem above. The second one which has the same core codebase somehow encounter that same error for other types.

``` ts
import type {
  EntityDTO,
  EntityKey,
  EntityMetadata,
  EntityName,
  FilterQuery,
  FindByCursorOptions,
  FindOneOptions,
  FindOptions,
  FromEntityType,
  GetRepository,
  Loaded,
  QueryOrderMap,
  RequiredEntityData,
  EnsureRequestContext,
  SqlEntityRepository,
  raw,
  wrap,
} from '@mikro-orm/postgresql';
```
All the types get this error:
``` ts
import { FindByCursorOptions, FindOneOptions, FindOptions, RequiredEntityData, } from '@mikro-orm/postgresql';
         ^^^^^^^^^^^^^^^^^^^
SyntaxError: The requested module '@mikro-orm/core' does not provide an export named 'FindByCursorOptions'
```

And I have to move them all to import type again so `pnpm run schema -d` could work. I don't know why this happens because it's literally the same codebase. Though this is also kind of an easy fix because it only happens once in my base service.
``` ts
import type {
  EntityDTO,
  EntityKey,
  EntityMetadata,
  EntityName,
  FilterQuery,
  FindByCursorOptions,
  FindOneOptions,
  FindOptions,
  FromEntityType,
  GetRepository,
  Loaded,
  QueryOrderMap,
  RequiredEntityData,
} from '@mikro-orm/postgresql';
import {
  EnsureRequestContext,
  SqlEntityRepository,
  raw,
  wrap,
} from '@mikro-orm/postgresql';
```

**Comment by linkthai** [comment]
After I upgraded to 6.0.2, I can import `Opt` directly. But it only works in `string & Opt` format and not `Opt<string>`. I consider this fixed, thanks a lot for your help. @B4nan

**Reply by B4nan** [reply]
Weird, nothing really changed in this direction, maybe you also updated some other dependencies on the way.

**Reply by linkthai** [reply]
I'm not sure about that, I use tilde for all my semver dependencies and I have already tried remove `pnpm-lock` and `node_modules` and reinstall 6.0.1 (and the previous rc version) and the `Opt` error still happened. It only stops happening on 6.0.2, which I read there were some changes regarding `Opt` and `Hidden` and decided to give it a try.

**Reply by B4nan** [reply]
The changes were in the ts-morph provider that was not detecting them correctly, nothing that should affect imports of the types.

Could be something about the paths mapping support (so external plugins). I also have a hunch that it could be about the ESM types, we don't have them and reuse the CJS type definitions. And of course, this could be as well caused by the custom loader (which uses the paths mapping, so could be a combination).
